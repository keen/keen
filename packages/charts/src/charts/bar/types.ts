export type Bar = {
  key: string;
  selector: (string | number)[];
  value: number;
  x: number;
  y: number;
  height: number;
  width: number;
  color: string;
  colorOutOfRange: boolean;
};
