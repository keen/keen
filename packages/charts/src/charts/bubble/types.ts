import { DataSelector } from '../../types';

export type Bubble = {
  key: string;
  x: number;
  y: number;
  radius: number;
  color: string;
  selector: DataSelector;
};
