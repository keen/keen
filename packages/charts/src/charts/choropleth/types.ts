import { Formatter } from '@keen.io/charts-utils';

export type Projection =
  | 'mercator'
  | 'geoAlbersUsa'
  | 'orthographic'
  | 'azimuthalEqualArea'
  | 'equalEarth'
  | 'naturalEarth';

export type ProjectionState = {
  translation: [number, number];
  scale: number;
  rotation: [number, number, number];
};

export type TooltipMeta = {
  color: string;
  label: string;
  value: number;
};

export type TooltipSettings = {
  formatValue?: Formatter;
};
