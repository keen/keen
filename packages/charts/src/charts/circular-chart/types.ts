import { Formatter } from '@keen.io/charts-utils';
import { Dimension, Margins, DataSelector } from '../../types';

export type LabelsPosition = 'inside' | 'outside';

export type SliceType = 'pie' | 'donut';

export type Options = {
  data: Record<string, any>[];
  labelSelector: string;
  keys: string[];
  disabledLabels: string[];
  dimension: Dimension;
  margins: Margins;
  colors: string[];
  padAngle: number;
  padRadius: number;
  cornerRadius: number;
  innerRadius: number;
  labelsRadius: number;
  labelsPosition: LabelsPosition;
  type?: SliceType;
  treshold?: number;
  formatValue?: Formatter;
  dataSeriesOffset?: [number, number];
};

export type Arc = {
  index: string;
  labelNumeric: string | number | Date | boolean;
  labelPercentage: string;
  activePosition: [number, number];
  labelPosition: [number, number];
  color: string;
  selector: DataSelector;
  dataKey: string;
  startAngle: number;
  endAngle: number;
  stacked: boolean;
  stack: { selector: DataSelector; color: string }[];
};

export type Slice = {
  color: string;
  value: number;
  dataKey: string;
  selector: DataSelector;
  stacked?: boolean;
  stack?: { selector: DataSelector; color: string }[];
  index?: number;
};

export type ArcProperties = {
  startAngle: number;
  endAngle: number;
};
