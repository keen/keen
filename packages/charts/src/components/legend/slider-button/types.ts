export type Variant = 'vertical' | 'horizontal';
