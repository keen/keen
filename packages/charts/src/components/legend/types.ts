export type LegendCardSettings = {
  backgroundColor?: string;
  border?: string;
  borderRadius?: string;
  hasShadow?: boolean;
  hideOverflow?: boolean;
  hasPadding?: boolean;
};
