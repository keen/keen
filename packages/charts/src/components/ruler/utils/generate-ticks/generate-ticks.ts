import { ScaleLinear, ScaleBand, ScaleTime } from 'd3-scale';
import {
  TimePrecision,
  ScaleSettings,
  getScaleTicks,
  getScaleCenterPosition,
} from '@keen.io/charts-utils';

import { Tick, Orientation } from '../../../../types';

const TICK_ALIGN = 0.5;

export const EDGE_TICK_ALIGN = 4;

export const generateTicks = ({
  scale,
  tickSize,
  orientation = Orientation.HORIZONTAL,
  scaleSettings,
  x,
  y,
}: {
  x: number;
  y: number;
  tickSize: number;
  scale:
    | ScaleBand<string>
    | ScaleLinear<number, number>
    | ScaleTime<number, number>;
  orientation?: Orientation;
  scaleSettings?: ScaleSettings;
}): {
  ticks: Tick[];
  ticksPrecision: TimePrecision;
} => {
  const { ticks: values, ticksPrecision } = getScaleTicks(scale, scaleSettings);

  const ticks: Tick[] = [];
  const position = 'bandwidth' in scale ? getScaleCenterPosition(scale) : scale;

  const getX = (value: string & { valueOf(): number }) =>
    orientation === Orientation.HORIZONTAL ? position(value) : x + TICK_ALIGN;

  const getY = (value: string & { valueOf(): number }) =>
    orientation === Orientation.HORIZONTAL ? y - TICK_ALIGN : position(value);

  values.forEach((value: any) => {
    ticks.push({
      size: tickSize,
      text: value,
      x: getX(value),
      y: getY(value),
    });
  });

  if ('bandwidth' in scale && orientation === Orientation.HORIZONTAL) {
    const [scaleStart, scaleEnd] = scale.range();

    ticks.push(
      {
        size: tickSize + EDGE_TICK_ALIGN,
        x: scaleStart,
        y: y - TICK_ALIGN,
      },
      {
        size: tickSize + EDGE_TICK_ALIGN,
        x: scaleEnd,
        y: y - TICK_ALIGN,
      }
    );
  }

  return {
    ticks,
    ticksPrecision,
  };
};

export default generateTicks;
