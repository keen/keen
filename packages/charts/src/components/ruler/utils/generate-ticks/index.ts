import generateTicks from './generate-ticks';

export default generateTicks;
