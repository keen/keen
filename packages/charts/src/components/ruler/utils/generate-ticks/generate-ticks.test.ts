import { scaleBand, scaleLinear, scaleUtc } from 'd3-scale';
import {
  register as registerTimezone,
  unregister as unregisterTimezone,
} from 'timezone-mock';
import { ScaleSettings } from '@keen.io/charts-utils';

beforeAll(() => {
  registerTimezone('UTC');
});

afterAll(() => {
  unregisterTimezone();
});

import { generateTicks, EDGE_TICK_ALIGN } from './generate-ticks';

import { Orientation } from '../../../../types';

test('create additional ticks for "band" scale with time precision', () => {
  const domain = ['Sales', 'Marketing', 'E-commerce'];
  const scaleSettings: ScaleSettings = {
    type: 'time',
    precision: 'month',
  };

  const scale = scaleBand()
    .range([
      new Date('2020-01-01T00:00:00.000Z'),
      new Date('2020-06-01T00:00:00.000Z'),
    ])
    .domain(domain);

  const { ticks, ticksPrecision } = generateTicks({
    x: 0,
    y: 0,
    scale,
    scaleSettings,
    tickSize: 10,
  });

  expect(ticks.length).toEqual(5);
  expect(ticksPrecision).toEqual(scaleSettings.precision);
});

test('creates ticks for vertical linear scale', () => {
  const scale = scaleLinear().range([0, 0]).domain([0, 90]);

  const ticks = generateTicks({
    x: 0,
    y: 0,
    scale,
    tickSize: 5,
    orientation: Orientation.VERTICAL,
  });

  expect(ticks).toMatchSnapshot();
});

test('creates ticks for horizontal linear scale', () => {
  const scale = scaleLinear().range([0, 0]).domain([0, 90]);

  const ticks = generateTicks({
    x: 0,
    y: 0,
    scale,
    tickSize: 5,
    orientation: Orientation.HORIZONTAL,
  });

  expect(ticks).toMatchSnapshot();
});

test('creates ticks for time scale', () => {
  const scale = scaleUtc()
    .range([0, 150])
    .domain([
      new Date('2020-01-01T00:00:00.000Z'),
      new Date('2020-06-01T00:00:00.000Z'),
    ]);

  const scaleSettings: ScaleSettings = {
    type: 'time',
    precision: 'month',
  };

  const ticks = generateTicks({
    x: 0,
    y: 0,
    scale,
    tickSize: 10,
    orientation: Orientation.VERTICAL,
    scaleSettings,
  });

  expect(ticks).toMatchInlineSnapshot(`
    Object {
      "ticks": Array [
        Object {
          "size": 10,
          "text": "2020-01-01T00:00:00.000Z",
          "x": 0.5,
          "y": 0,
        },
        Object {
          "size": 10,
          "text": "2020-04-01T00:00:00.000Z",
          "x": 0.5,
          "y": 89.80263157894737,
        },
      ],
      "ticksPrecision": "month",
    }
  `);
});

test('increase size of scale "band" edge ticks', () => {
  const domain = ['Sales', 'Marketing', 'E-commerce'];
  const tickSize = 5;

  const scale = scaleBand()
    .range([
      new Date('2020-01-01T00:00:00.000Z'),
      new Date('2020-06-01T00:00:00.000Z'),
    ])
    .domain(domain);

  const { ticks } = generateTicks({
    x: 0,
    y: 0,
    scale,
    tickSize,
  });

  const edgeTicks = ticks.slice(ticks.length - 2, ticks.length);
  edgeTicks.forEach(({ size }) => {
    expect(size).toEqual(tickSize + EDGE_TICK_ALIGN);
  });
});
