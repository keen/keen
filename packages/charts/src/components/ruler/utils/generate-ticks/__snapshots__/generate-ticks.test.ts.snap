// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates ticks for horizontal linear scale 1`] = `
Object {
  "ticks": Array [
    Object {
      "size": 5,
      "text": 0,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 10,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 20,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 30,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 40,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 50,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 60,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 70,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 80,
      "x": 0,
      "y": -0.5,
    },
    Object {
      "size": 5,
      "text": 90,
      "x": 0,
      "y": -0.5,
    },
  ],
  "ticksPrecision": undefined,
}
`;

exports[`creates ticks for vertical linear scale 1`] = `
Object {
  "ticks": Array [
    Object {
      "size": 5,
      "text": 0,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 10,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 20,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 30,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 40,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 50,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 60,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 70,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 80,
      "x": 0.5,
      "y": 0,
    },
    Object {
      "size": 5,
      "text": 90,
      "x": 0.5,
      "y": 0,
    },
  ],
  "ticksPrecision": undefined,
}
`;
